### PR TITLE
Write `rustfmt` safety manual

### PR DESCRIPTION
**Pull request by Urhengulas**
Do the constraints and tool option make sense?

### CODE REVIEW

File: ferrocene/doc/safety-manual/src/rustfmt/options.rst [hunk 1]
```diff
@@ -0,0 +1,15 @@
+.. SPDX-License-Identifier: MIT OR Apache-2.0
+   SPDX-FileCopyrightText: The Ferrocene Developers
+
+Tool Options
+============
+
+``rustfmt`` is qualified exclusively for the following command line options:
+
+* Users shall pass command line option ``--emit files`` to each invocation of ``rustfmt``.
```
Comment (pietroalbini) [inline ferrocene/doc/safety-manual/src/rustfmt/options.rst:9]: Why are we forcing people to pass `--emit files`?

File: ferrocene/doc/safety-manual/src/rustfmt/options.rst [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+* Users shall pass command line option ``--edition 2021`` to each invocation of ``rustfmt``.
```
Comment (pietroalbini) [inline ferrocene/doc/safety-manual/src/rustfmt/options.rst:10]: Why are we limiting to just `--edition 2021`? We do so for rustc as we only specified the 2021 edition in the FLS, but there is no such constraint for rustfmt.

File: ferrocene/doc/safety-manual/src/rustfmt/constraints.rst [hunk 3]
```diff
@@ -0,0 +1,15 @@
+.. SPDX-License-Identifier: MIT OR Apache-2.0
+   SPDX-FileCopyrightText: The Ferrocene Developers
+
+Constraints
```
Comment (pietroalbini) [inline ferrocene/doc/safety-manual/src/rustfmt/constraints.rst:4]: Constraints should be styled like rustc constraints: a section for each constraint, with a unique ID for each constraint and linking to the nodes of the HazOp they address.

File: ferrocene/doc/safety-manual/src/rustfmt/options.rst [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+.. |Tool| replace:: ``rustfmt``
+.. |Link| replace:: :doc:`user-manual:rustfmt/cli`
+.. include:: /partials/options.rst
```
Comment (pietroalbini) [inline ferrocene/doc/safety-manual/src/rustfmt/options.rst:14]: I would not share the snippet across the various tools.

A lot of the text in the snippet for rustc doesn't apply: there are no flags that we restrict in the documentation, I don't think we need to mandate people pass any flags all the time, and these flags don't affect codegen. I think we can get away with a far smaller piece of text here.

Also, this approach for partials is pretty hacky, and I'd prefer a better templating solution if we were to need to share snippets in the future. Let's discuss about that in Zulip.

**Review by pietroalbini — CHANGES_REQUESTED** [review]

**Comment by Urhengulas** [comment]
Blocked on #666

File: ferrocene/doc/safety-manual/src/rustfmt/constraints.rst [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+Install version control system
+------------------------------
+
+.. id:: RUSTFMT_CSTR_0010_HAVE_VCS
+
+Associated requirement ID: :id:`RUSTFMT_AVD_CHECK_DIFF_001`
+
+Before executing ``rustfmt``, the user shall ensure that the a version control
+system (VCS, e.g. ``git``) is installed and used.
```
Comment (pietroalbini) [inline ferrocene/doc/safety-manual/src/rustfmt/constraints.rst:17]: If we want to require a VCS (see next comment), having it installed shouldn't be a constraint, it should be a pre-requisite in the installation process.

File: ferrocene/doc/safety-manual/src/rustfmt/constraints.rst [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+Review diff
+-----------
+
+.. id:: RUSTFMT_CSTR_0020_USE_VCS
+
+Associated requirement ID: :id:`RUSTFMT_AVD_CHECK_DIFF_001`
+
+After executing ``rustfmt``, the user shall review, with the help of the VCS, the
+modifications ``rustfmt`` performed and ensure that there are no semantic changes.
```
Comment (pietroalbini) [inline ferrocene/doc/safety-manual/src/rustfmt/constraints.rst:27]: Should we mandate the use of a VCS at all? I would hope that our customers use one, but if they want to use a different technique to review the changes they should be free to.

File: ferrocene/doc/safety-manual/src/rustfmt/constraints.rst [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+sufficient tests that ensure the correctness of the code.
+
+Tests pass
+----------
+
+.. id:: RUSTFMT_CSTR_0030_USE_TESTS
+
+Associated requirement ID: :id:`RUSTFMT_AVD_TESTS_002`
+
+After executing ``rustfmt``, the user shall ensure that all tests pass.
```
Comment (pietroalbini) [inline ferrocene/doc/safety-manual/src/rustfmt/constraints.rst:46]: We should probably merge the two requirements into one.

**Review by pietroalbini — CHANGES_REQUESTED** [review]

File: ferrocene/doc/safety-manual/src/rustfmt/constraints.rst (hunk 6 above)
Comment (tshepang) [reply ferrocene/doc/safety-manual/src/rustfmt/constraints.rst:27]: I think we should not mandate VCS, especially since `rustfmt` works fine without it

Comment (pietroalbini) [reply ferrocene/doc/safety-manual/src/rustfmt/constraints.rst:27]: Like, we do want to mandate that they manually review the changes done by rustfmt did not cause any difference in the semantic meaning of the code, but we shouldn't say *how* they do it.

File: ferrocene/doc/safety-manual/src/rustfmt/constraints.rst [hunk 8]
```diff
@@ -6,30 +6,22 @@
...
+.. id:: RUSTFMT_CSTR_0010_CHECK_DIFF
 
 Associated requirement ID: :id:`RUSTFMT_AVD_CHECK_DIFF_001`
 
-After executing ``rustfmt``, the user shall review, with the help of the VCS, the
-modifications ``rustfmt`` performed and ensure that there are no semantic changes.
+After executing ``rustfmt``, the user shall review the modifications
+``rustfmt`` performed and ensure that there are no semantic changes.
+
+Using a version control system for that makes sense, but is not mandated.
```
Comment (tshepang) [inline ferrocene/doc/safety-manual/src/rustfmt/constraints.rst:19]: would not mention version control at all... users can decide what *makes sense* for them

**Review by tshepang — APPROVED** [review]
bors merge

**Comment by tshepang** [comment]
bors retry

**Pull request merged**